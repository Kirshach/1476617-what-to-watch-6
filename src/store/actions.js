export const ActionType = {
  SET_GENRE: `films/setGenre`,
};

export const setGenre = (payload) => ({
  type: ActionType.SET_GENRE,
  payload
});
