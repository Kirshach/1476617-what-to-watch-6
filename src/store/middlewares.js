import {history} from '../history';

export const redirectActionType = `app/redirect`;

export const redirectAction = (payload) => ({
  type: redirectActionType,
  payload,
});

export const redirectMiddleware = (_store) => (next) => (action) => {
  if (action.type === redirectActionType) {
    history.push(action.payload);
  }
  return next(action);
};
