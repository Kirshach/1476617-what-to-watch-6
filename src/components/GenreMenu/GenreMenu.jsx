import React from 'react';
import {useSelector, useDispatch} from 'react-redux';

import {setGenreAction, showDefaultFilmsAmountAction} from '../../store/app/state/actions';
import {genreSelector} from '../../store/app/state/selectors';
import {Genres} from '../../const';

const genres = Object.values(Genres);

export const GenreMenu = () => {
  const dispatch = useDispatch();
  const currentGenre = useSelector(genreSelector);

  const getClassName = (genre) => `catalog__genres-item ${currentGenre === genre ? `catalog__genres-item--active` : ``}`;
  const getHandleGenreClickHandler = (genre) => (evt) => {
    evt.preventDefault();
    dispatch(setGenreAction(genre));
    dispatch(showDefaultFilmsAmountAction());
  };

  return (
    <ul className="catalog__genres-list">
      {genres.map((genre) => (
        <li key={genre} className={getClassName(genre)}>
          <a href="#" onClick={getHandleGenreClickHandler(genre)} className="catalog__genres-link">
            {genre}
          </a>
        </li>
      ))}
    </ul>
  );
};

export default GenreMenu;
