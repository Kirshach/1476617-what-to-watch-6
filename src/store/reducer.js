import {ActionType} from './actions';
import {Genres} from '../const';
import {films} from '../mocks/films';


const initialState = {
  films,
  genre: Genres.allGenres,
  genres: Object.values(Genres),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_GENRE:
      return {
        ...state,
        genre: action.payload
      };
    default:
      return state;
  }
};

export {reducer};
