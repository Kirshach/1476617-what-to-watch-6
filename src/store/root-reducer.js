import {combineReducers} from 'redux';

import {appReducer} from './app/reducer';
import {domainReducer} from './domain/reducer';

export const DOMAIN = `domain`;
export const APP = `app`;

const rootReducer = combineReducers({
  [DOMAIN]: domainReducer,
  [APP]: appReducer,
});

export {rootReducer};

export const RootNS = {DOMAIN, APP};
