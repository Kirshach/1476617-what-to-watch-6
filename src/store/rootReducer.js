import {combineReducers} from 'redux';

import {appReducer} from './app/reducer';
import {domainReducer} from './domain/reducer';

const DOMAIN = `domain`;
const APP = `app`;

const rootReducer = combineReducers({
  [DOMAIN]: domainReducer,
  [APP]: appReducer,
});

export {rootReducer};

export const RootNS = {DOMAIN, APP};
