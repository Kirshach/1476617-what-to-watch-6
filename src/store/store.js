import {createStore, applyMiddleware} from 'redux';
import {composeWithDevTools} from 'redux-devtools-extension';
import thunk from 'redux-thunk';

import {redirectMiddleware, handleAPIErrorMiddleware} from './middlewares';
import {checkAuthThunk} from '../store/app/auth/thunks';
import {fetchFilmsThunk} from '../store/domain/thunks';
import {rootReducer as reducer} from '../store/rootReducer';

import {api} from '../api/api';

const store = createStore(
    reducer,
    composeWithDevTools(
        applyMiddleware(
            handleAPIErrorMiddleware,
            redirectMiddleware,
            thunk.withExtraArgument(api),
        ),
    ),
);

store.dispatch(checkAuthThunk());
store.dispatch(fetchFilmsThunk());

export {store};
