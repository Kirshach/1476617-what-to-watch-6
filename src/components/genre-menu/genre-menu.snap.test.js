import React from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {render} from '@testing-library/react';

import {store} from '../../mocks/store/store';

import GenreMenu from './genre-menu';

jest.mock(`react-redux`);

test(`<GenreMenu /> renders correctly`, () => {
  useDispatch.mockReturnValue(() => Promise.resolve());
  useSelector.mockImplementation((selector) => selector(store));

  const {container} = render(<GenreMenu onGenreChange={() => {}} />);
  expect(container).toMatchSnapshot();
});
