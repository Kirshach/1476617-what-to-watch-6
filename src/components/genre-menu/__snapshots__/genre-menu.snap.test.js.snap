// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GenreMenu /> renders correctly 1`] = `
<div>
  <ul
    class="catalog__genres-list"
  >
    <li
      class="catalog__genres-item catalog__genres-item--active"
    >
      <a
        class="catalog__genres-link"
        href="#"
      >
        All genres
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="#"
      >
        Action
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="#"
      >
        Adventure
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="#"
      >
        Comedy
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="#"
      >
        Crime
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="#"
      >
        Drama
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="#"
      >
        Fantasy
      </a>
    </li>
    <li
      class="catalog__genres-item "
    >
      <a
        class="catalog__genres-link"
        href="#"
      >
        Thriller
      </a>
    </li>
  </ul>
</div>
`;
